ios/Runner/AppDelegate.swift

```diff
@@ -3,11 +3,64 @@ import Flutter
 
 @main
 @objc class AppDelegate: FlutterAppDelegate {
+  private var notificationTapEventListener: NotificationTapEventListener?
+
   override func application(
     _ application: UIApplication,
     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
   ) -> Bool {
     GeneratedPluginRegistrant.register(with: self)
+    let controller = window?.rootViewController as! FlutterViewController
+
+    // Retrieve the remote notification payload from launch options;
+    // this will be null if the launch wasn't triggered by a notification.
+    let notificationPayload = launchOptions?[.remoteNotification] as? [AnyHashable : Any]
+    let api = NotificationHostApiImpl(notificationPayload.map { NotificationDataFromLaunch(payload: $0) })
+    NotificationHostApiSetup.setUp(binaryMessenger: controller.binaryMessenger, api: api)
+
+    notificationTapEventListener = NotificationTapEventListener()
+    NotificationTapEventsStreamHandler.register(with: controller.binaryMessenger, streamHandler: notificationTapEventListener!)
+
+    // Setup handler for notification tap while the app is running.
+    UNUserNotificationCenter.current().delegate = self
+
     return super.application(application, didFinishLaunchingWithOptions: launchOptions)
   }
+
+  override func userNotificationCenter(
+    _ center: UNUserNotificationCenter,
+    didReceive response: UNNotificationResponse,
+    withCompletionHandler completionHandler: @escaping () -> Void
+  ) {
+    if response.actionIdentifier == UNNotificationDefaultActionIdentifier {
+      let listener = notificationTapEventListener!
+      let userInfo = response.notification.request.content.userInfo
+      listener.onNotificationTapEvent(payload: userInfo)
+    }
+    completionHandler()
+  }
+}
+
+private class NotificationHostApiImpl: NotificationHostApi {
+  private let maybeDataFromLaunch: NotificationDataFromLaunch?
+
+  init(_ maybeDataFromLaunch: NotificationDataFromLaunch?) {
+    self.maybeDataFromLaunch = maybeDataFromLaunch
+  }
+
+  func getNotificationDataFromLaunch() -> NotificationDataFromLaunch? {
+    maybeDataFromLaunch
+  }
+}
+
+class NotificationTapEventListener: NotificationTapEventsStreamHandler {
```

Comment:
Let's include comments somewhere in here with links to the Flutter examples that you found helpful, from the thread https://github.com/zulip/zulip-flutter/pull/1379#discussion_r2027866613 , since those examples were nontrivial to find.

Here at the top of this class is probably a fine place.